core/retries/src/main/java/software/amazon/awssdk/retries/internal/AdaptiveRetryStrategyImpl.java

```diff
@@ -0,0 +1,374 @@
+                     .build();
+            throw new TokenAcquisitionFailedException(message, refreshedToken, failure);
+        }
+        int attempt = token.attempt();
+        LOG.warn(() -> String.format("Request attempt %d encountered retryable failure.", attempt), failure);
+    }
+
+    private String nonRetryableExceptionMessage(DefaultRetryToken token) {
+        return String.format("Request attempt %d encountered non-retryable failure", token.attempt());
+    }
+
+    private String maxAttemptsReachedMessage(DefaultRetryToken token) {
+        return String.format("Request will not be retried. Retries have been exhausted "
+                             + "(cost: 0, capacity: %d/%d)",
+                             token.capacityAcquired(),
+                             token.capacityRemaining());
+    }
+
+    private String acquisitionFailedMessage(AcquireResponse acquireResponse) {
+        return String.format("Request will not be retried to protect the caller and downstream service. "
+                             + "The cost of retrying (%d) "
+                             + "exceeds the available retry capacity (%d/%d).",
+                             acquireResponse.capacityRequested(),
+                             acquireResponse.capacityRemaining(),
+                             acquireResponse.maxCapacity());
+    }
+
+    private void logAcquireInitialToken(AcquireInitialTokenRequest request) {
+        // Request attempt 1 token acquired (backoff: 0ms, cost: 0, capacity: 500/500)
+        TokenBucket tokenBucket = tokenBucketStore.tokenBucketForScope(request.scope());
+        LOG.debug(() -> String.format("Request attempt 1 token acquired "
+                                      + "(backoff: 0ms, cost: 0, capacity: %d/%d)",
+                                      tokenBucket.currentCapacity(), tokenBucket.maxCapacity()));
+    }
+
+    private void logRefreshTokenSuccess(DefaultRetryToken token, AcquireResponse acquireResponse, Duration delay) {
+        LOG.debug(() -> String.format("Request attempt %d token acquired "
+                                      + "(backoff: %dms, cost: %d, capacity: %d/%d)",
+                                      token.attempt(), delay.toMillis(),
+                                      acquireResponse.capacityAcquired(),
+                                      acquireResponse.capacityRemaining(),
+                                      acquireResponse.maxCapacity()));
+    }
+
+    private void logRecordSuccess(DefaultRetryToken token, ReleaseResponse release) {
+        LOG.debug(() -> String.format("Request attempt %d succeeded (cost: -%d, capacity: %d/%d)",
+                                      token.attempt(), release.capacityReleased(),
+                                      release.currentCapacity(), release.maxCapacity()));
+
+    }
+
+    private boolean maxAttemptsReached(DefaultRetryToken token) {
+        return token.attempt() >= maxAttempts;
```

Comment:
No, the class returns `int` instead of `Integer`. See the definition [here](https://github.com/aws/aws-sdk-java-v2/blob/feature/master/sra-retries/core/retries/src/main/java/software/amazon/awssdk/retries/internal/DefaultRetryToken.java#L53).